sys/kern/lockdep.c

```diff
@@ -0,0 +1,257 @@
+  spin_unlock(&main_lock);
+}
+
+static inline lock_class_link_t *bfs_next_link(lock_class_link_t *link) {
+  if (link == NULL)
+    return NULL;
+  return SIMPLEQ_NEXT(link, entry);
+}
+
+static inline void bfs_mark_visited(lock_class_t *class) {
+  class->bfs_gen_id = bfs_cur_gen;
+}
+
+static inline int bfs_is_visited(lock_class_t *class) {
+  return class->bfs_gen_id == bfs_cur_gen;
+}
+
+/*
+ * Check that the dependency graph starting at <src> can lead to
+ * <target> or not.
+ */
+static bool check_path(lock_class_t *src, lock_class_t *target) {
+  lock_class_link_t *child, *link = NULL;
+  bfs_queue_t *queue = &bfs_queue;
+
+  bfs_q_init(queue);
+  bfs_cur_gen++;
+
+  bfs_q_enqueue(queue, SIMPLEQ_FIRST(&src->locked_after));
+  bfs_mark_visited(src);
+
+  while ((link = bfs_next_link(link)) || (link = bfs_q_dequeue(queue))) {
+    if (bfs_is_visited(link->to))
+      continue;
+    bfs_mark_visited(link->to);
+
+    if (link->to == target)
+      return 1;
+
+    if ((child = SIMPLEQ_FIRST(&link->to->locked_after))) {
+      if (!bfs_q_enqueue(queue, child))
+        panic("lockdep: bfs queue overflow");
+    }
+  }
+
+  return 0;
+}
+
+static lock_class_t *look_up_lock_class(lock_class_mapping_t *lock) {
+  lock_class_key_t *key = lock->key;
+  lock_class_t *class;
+
+  SIMPLEQ_FOREACH(class, CLASS_HASH_CHAIN(key), hash_entry) {
+    if (class->key == key)
+      return class;
+  }
+
+  return NULL;
+}
+
+static lock_class_t *alloc_class(void) {
+  if (class_cnt >= MAX_CLASSES)
+    panic("lockdep: no more classes");
+
+  return &lock_classes[class_cnt++];
+}
+
+static lock_class_t *get_or_create_class(lock_class_mapping_t *lock) {
+  lock_class_t *class;
+
+  /* If the lock doesn't have a key then it is statically allocated. In this
+   * case use its address as the key. */
+  if (lock->key == NULL)
+    lock->key = (void *)lock;
+
+  class = look_up_lock_class(lock);
+  if (!class) {
+    class = alloc_class();
+    class->key = lock->key;
+    class->name = lock->name;
+    SIMPLEQ_INIT(&class->locked_after);
+    SIMPLEQ_INSERT_HEAD(CLASS_HASH_CHAIN(lock->key), class, hash_entry);
+  }
+  lock->lock_class = class;
+
+  return class;
+}
+
+static lock_class_link_t *alloc_link(void) {
+  if (link_cnt >= MAX_LINKS)
+    panic("lockdep: no more links");
+
+  return &lock_links[link_cnt++];
+}
+
+static void add_prev_link(void) {
+  held_lock_t *hprev, *hcur;
+  lock_class_link_t *link;
+  int depth = thread_self()->td_lock_depth;
+
+  if (depth < 2)
+    return;
+
+  hcur = &thread_self()->td_held_locks[depth - 1];
+  hprev = &thread_self()->td_held_locks[depth - 2];
+
+  if (hcur->lock_class == hprev->lock_class)
+    return;
+
+  SIMPLEQ_FOREACH(link, &(hprev->lock_class->locked_after), entry) {
+    if (hcur->lock_class == link->to)
+      return;
+  }
+
+  link = alloc_link();
+  link->to = hcur->lock_class;
+
+  SIMPLEQ_INSERT_HEAD(&(hprev->lock_class->locked_after), link, entry);
+
+  if (check_path(link->to, hprev->lock_class)) {
```

Comment:
```suggestion
  if (check_path(hcur->lock_class, hprev->lock_class)) {
```
I think it's more readable that way